packages/eslint-plugin-react-hooks/src/shared/RunReactCompiler.ts

```diff
@@ -0,0 +1,303 @@
+      });
+    } catch {
+      /* empty */
+    }
+  }
+
+  if (babelAST != null) {
+    results.flowSuppressions = getFlowSuppressions(sourceCode);
+    try {
+      transformFromAstSync(babelAST, sourceCode.text, {
+        filename,
+        highlightCode: false,
+        retainLines: true,
+        plugins: [
+          [PluginProposalPrivateMethods, {loose: true}],
+          [BabelPluginReactCompiler, options],
+        ],
+        sourceType: 'module',
+        configFile: false,
+        babelrc: false,
+      });
+
+      if (results.events.filter(e => e.kind === 'CompileError').length === 0) {
+        traverse(babelAST, {
+          FunctionDeclaration(path) {
+            results.unusedOptOutDirectives.push(
+              ...filterUnusedOptOutDirectives(path.node.body.directives),
+            );
+          },
+          ArrowFunctionExpression(path) {
+            if (path.node.body.type === 'BlockStatement') {
+              results.unusedOptOutDirectives.push(
+                ...filterUnusedOptOutDirectives(path.node.body.directives),
+              );
+            }
+          },
+          FunctionExpression(path) {
+            results.unusedOptOutDirectives.push(
+              ...filterUnusedOptOutDirectives(path.node.body.directives),
+            );
+          },
+        });
+      }
+    } catch (err) {
+      /* errors handled by injected logger */
+    }
+  }
+
+  return results;
+}
+
+const SENTINEL = Symbol();
+
+// Array backed LRU cache -- should be small < 10 elements
+class LRUCache<K, T> {
+  // newest at headIdx, then headIdx + 1, ..., tailIdx
+  #values: Array<[K, T | Error] | [typeof SENTINEL, void]>;
+  #headIdx: number = 0;
+
+  constructor(size: number) {
+    this.#values = new Array(size).fill(SENTINEL);
+  }
+
+  // gets a value and sets it as "recently used"
+  get(key: K): T | null {
+    const idx = this.#values.findIndex(entry => entry[0] === key);
+    // If found, move to front
+    if (idx === this.#headIdx) {
+      return this.#values[this.#headIdx][1] as T;
+    } else if (idx < 0) {
+      return null;
+    }
+
+    const entry: [K, T] = this.#values[idx] as [K, T];
+
+    const len = this.#values.length;
+    for (let i = 0; i < Math.min(idx, len - 1); i++) {
+      this.#values[(this.#headIdx + i + 1) % len] =
+        this.#values[(this.#headIdx + i) % len];
+    }
+    this.#values[this.#headIdx] = entry;
+    return entry[1];
+  }
+  push(key: K, value: T): void {
+    this.#headIdx =
+      (this.#headIdx - 1 + this.#values.length) % this.#values.length;
+    this.#values[this.#headIdx] = [key, value];
+  }
+}
+const cache = new LRUCache<string, RunCacheEntry>(10);
+
+export default function runReactCompiler({
+  sourceCode,
+  filename,
+  userOpts,
+}: RunParams): RunCacheEntry {
+  const entry = cache.get(filename);
+  if (
+    entry != null &&
+    entry.sourceCode === sourceCode.text &&
+    isDeepStrictEqual(entry.userOpts, userOpts)
+  ) {
+    return entry;
```

Comment:
just curious, can we include a test plan in the PR showing that we do get cache hits?